# Interactuando con una base de datos relacional utilizando SQL

Como Ingeniero de Datos, quieres ayudar a la organización a encontrar información valiosa a través de los datos. Para hacerlo, debes aprender los conceptos básicos de consultar y manipular datos con SQL.

En este laboratorio, trabajarás con una base de datos relacional de muestra de alquiler de DVD donde encontrarás información sobre tiendas, clientes y películas alquiladas. A continuación, encontrarás el diagrama que muestra las relaciones entre las tablas y una breve descripción de cada tabla.

El propósito de este laboratorio es utilizar el lenguaje SQL para consultar la base de datos con el fin de responder preguntas comerciales. A medida que completes cada ejercicio, debes verificar la salida que obtienes de cada bloque de celdas de ejercicio con la salida esperada. Si te quedas atascado, siempre puedes consultar el cuaderno de soluciones.

Para abrir el cuaderno de soluciones, sigue estos pasos:
- Ve al menú principal y selecciona `Archivo -> Preferencias -> Configuración`.
- Haz clic en `editor de texto` a la izquierda, luego desplázate hacia abajo hasta la sección `Excluir archivos`.
- Elimina la línea `**/C2_W1_Lab_1_SQL_Solution.ipynb`. El archivo ahora aparecerá en el explorador.
- Puedes cerrar la pestaña `Configuración`.


# Tabla de Contenidos
- [ 1 - Base de Datos](#1)
- [ 2 - Ejecución de Comandos SQL en un Cuaderno](#2)
- [ 3 - Operaciones de Crear, Leer, Actualizar y Eliminar (CRUD)](#3)
  - [ 3.1 - CREAR TABLA](#3-1)
    - [ Ejercicio 1](#ex01)
  - [ 3.2 - SELECT](#3-2)
    - [ Ejercicio 2](#ex02)
    - [ Ejercicio 3](#ex03)
  - [ 3.3 - WHERE](#3-3)
    - [ Ejercicio 4](#ex04)
  - [ 3.4 - INSERT INTO](#3-4)
    - [ Ejercicio 5](#ex05)
  - [ 3.5 - UPDATE](#3-5)
    - [ Ejercicio 6](#ex06)
  - [ 3.6 - DELETE](#3-6)
    - [ Ejercicio 7](#ex07)
- [ 4 - Cláusulas SQL](#4)
  - [ 4.1 - ALIAS](#4-1)
    - [ Ejercicio 8](#ex08)
  - [ 4.2 - JOIN](#4-2)
    - [ Ejercicio 9](#ex09)
    - [ Ejercicio 10](#ex10)
  - [ 4.3 - GROUP BY](#4-3)
    - [ Ejercicio 11](#ex11)
  - [ 4.4 - ORDER BY](#4-4)
    - [ Ejercicio 12](#ex12)


<a id='1'></a>
## 1 - Base de datos

Trabajarás con una versión modificada de la [Base de datos de muestra Sakila](https://dev.mysql.com/doc/sakila/en/), que está bajo la licencia [Nueva licencia BSD](https://dev.mysql.com/doc/sakila/en/sakila-license.html).

Para fines de aprendizaje, vamos a asumir que los datos pertenecen a _Rentio_, que es una empresa ficticia dedicada al alquiler de películas a clientes de todo el mundo.

En la base de datos, podrás encontrar los datos de las tiendas y el personal que trabaja en ellas, así como sus direcciones. Cada tienda gestiona su inventario, por lo que cuando una tienda recibe un nuevo DVD, se inserta información sobre la película, categoría, idioma y actores en la base de datos. Además, cada vez que un nuevo cliente alquila una película, se inserta la información básica del cliente en la base de datos junto con su dirección. Además, se agrega un alquiler tan pronto como se produce una transacción con información sobre inventario, película y clientes que pagan.

A lo largo de los ejercicios de este laboratorio, trabajarás con las diversas tablas de esta base de datos, ayudando a la empresa a comprender las películas más alquiladas por diferentes atributos.

La base de datos transaccional de Rentio incluye las siguientes tablas.

- `actor`: Contiene los datos del actor, como nombre y apellido.
- `address`: Contiene datos de dirección del personal y clientes.
- `category`: Contiene datos de categoría de la película.
- `city`: Tiene nombres de ciudades.
- `country`: Tiene nombres de países.
- `customer`: Contiene datos del cliente, como nombre, apellido, tiendas donde compraron y si está activo o no.
- `film`: Contiene datos de la película, como título, descripción, idioma y calificaciones.
- `film_actor`: Almacena la relación entre película y actor.
- `film_category`: Almacena la relación entre película y categoría.
- `inventory`: Contiene datos de inventario relacionados con las películas y la tienda donde se almacenan.
- `language`: Tiene nombres de idiomas.
- `payment`: Contiene datos de pago de clientes relacionados con el personal, los montos y las fechas.
- `rental`: Contiene datos de alquiler relacionados con el cliente, el personal, las fechas de alquiler y la fecha de devolución.
- `staff`: Contiene datos del personal, como nombre, apellido, tiendas donde trabajan y si están activos o no.
- `store`: Contiene datos de la tienda, como el gerente y la dirección de la tienda.

Aquí puedes encontrar el modelo entidad-relación (ERM) de la base de datos transaccional que muestra todas las tablas y cómo están relacionadas:

![rentio-database-erd](images/rentio_database_erd.png)


**Nota**:
Mientras vas bajando para hacer los ejercicios de SQL, puedes dividir este editor en dos para mantener el diagrama de la base de datos junto a ti.
- Escribe "Ctrl + \\" o "Cmd + \\", una nueva ventana debería abrirse a la derecha.
- Luego, desde el menú del explorador a la izquierda, puedes abrir la carpeta de imágenes y arrastrar la imagen rentio_database_erd_2.png a la segunda ventana. Y en esta ventana actual, puedes mantener abierto este cuaderno para completar los ejercicios.
- Siéntete libre de personalizar tus ventanas usando las opciones de alternancia que se encuentran en la esquina superior derecha de este editor.


<a id='2'></a>
## 2 - Ejecución de comandos SQL en un cuaderno

En este laboratorio, aprovecharás el uso de la "magia" SQL (`%sql` o `%%sql`) ofrecida por la [extensión](https://pypi.org/project/ipython-sql/) `ipython-sql`, que te permite conectarte a una base de datos y emitir comandos SQL de forma interactiva dentro de un cuaderno de JupyterLab. JupyterLab define "magia" como comandos especiales precedidos por `%`. Aquí, emplearás la magia `load_ext` para cargar la extensión `ipython-sql`. Carga el módulo SQL:


```bash
pip install ipython-sql
pip install sqlachemy
pip install pymysql
```


In [3]:
%load_ext sql

El comando mágico proporcionado carga la extensión `ipython-sql`, permitiendo la conexión a bases de datos compatibles con [SQLAlchemy](https://www.sqlalchemy.org/features.html). Antes de comenzar a emitir consultas SQL a la base de datos, primero debes establecer una conexión con la base de datos MYSQL. Para ello, necesitas información como el nombre de usuario de la base de datos, la contraseña, el nombre de host (o dirección), el número de puerto y el nombre de la base de datos. Esta información se te proporciona y se almacena en las variables de entorno definidas en el archivo `src/env`. Ejecuta la siguiente celda para cargar los valores de estas variables y establecer una conexión con la base de datos MYSQL.


In [4]:
import os 
# import socket

from dotenv import load_dotenv

load_dotenv('./src/env', override=True)

# DBHOST = socket.gethostname()
DBHOST = "127.0.0.1"
DBPORT = os.getenv('DBPORT')
DBNAME = os.getenv('DBNAME')
DBUSER = os.getenv('DBUSER')
DBPASSWORD = os.getenv('DBPASSWORD')

connection_url = f'mysql+pymysql://{DBUSER}:{DBPASSWORD}@{DBHOST}:{DBPORT}/{DBNAME}'

%sql {connection_url}

In [5]:
%sql

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila


In [6]:
%sql SHOW TABLES;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
23 rows affected.


[('actor',),
 ('actor_info',),
 ('address',),
 ('category',),
 ('city',),
 ('country',),
 ('customer',),
 ('customer_list',),
 ('film',),
 ('film_actor',),
 ('film_category',),
 ('film_list',),
 ('film_text',),
 ('inventory',),
 ('language',),
 ('nicer_but_slower_film_list',),
 ('payment',),
 ('rental',),
 ('sales_by_film_category',),
 ('sales_by_store',),
 ('staff',),
 ('staff_list',),
 ('store',)]

<a id='3'></a>
## 3 - Operaciones de Crear, Leer, Actualizar y Borrar (CRUD)

CRUD significa Crear, Leer, Actualizar y Borrar, que son operaciones básicas para manipular datos. Cuando hablamos de bases de datos, usamos las declaraciones `INSERT INTO`, `SELECT`, `UPDATE` y `DELETE` respectivamente para referirnos a las operaciones CRUD.


<a id='3-1'></a>
### 3.1 - CREAR TABLA

Antes de utilizar las declaraciones para operaciones CRUD, verás la declaración `CREATE TABLE` que se utiliza para crear una nueva tabla en una base de datos. Debes especificar el nombre de las columnas y el tipo de datos para cada columna. Puedes consultar la lista completa de tipos de datos [aquí](https://dev.mysql.com/doc/refman/8.0/en/data-types.html).

```sql
CREATE TABLE nombre_tabla (
    columna1 tipo_de_dato,
    columna2 tipo_de_dato,
    columna3 tipo_de_dato,
   ...
);
```


<a id='ex01'></a>
### Ejercicio 1

Escribe una consulta SQL para crear una réplica de la tabla `category` llamada `category_copy`. Utiliza estas columnas:

| nombre de la columna | tipo de dato |
| -------------------- | ------------ |
| category_id          | INTEGER      |
| name                 | VARCHAR(25)  |
| last_update          | TIMESTAMP    |

VARCHAR(25) significa una cadena de longitud de caracteres variable que puede contener hasta 25 caracteres en ella.


**Nota:** Estarás utilizando el comando mágico `%%sql` para permitir la sintaxis de consulta multilínea.

**Nota:** Reemplaza todos los `None` en el código proporcionado.


In [7]:
%%sql 
CREATE TABLE category_copy (
    category_id INTEGER,
    name VARCHAR(25),
    last_update TIMESTAMP
);


 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
0 rows affected.


[]

<a id='3-2'></a>
### 3.2 - SELECT

La instrucción `SELECT` se utiliza para obtener datos de una base de datos. También se combina con la cláusula `FROM` para indicar la tabla que deseas consultar.

Puedes especificar las columnas de la tabla que deseas recuperar de la consulta enumerando cada una de la siguiente manera:

```sql
SELECT
    columna1,
    columna2,
    columna3,
    ...
FROM nombre_tabla;
```

Además, puedes usar `*` para obtener todas las columnas de la tabla:

```sql
SELECT
    *
FROM nombre_tabla;
```

La cláusula `LIMIT` se utiliza para limitar la cantidad de filas que devuelve la consulta.

```sql
SELECT
    *
FROM tabla1
LIMIT 1;
```


<a id='ex02'></a>
### Ejercicio 2

Escribe una consulta SQL para recuperar el título, duración y año de lanzamiento de la tabla `film`. Limita la salida a solo 4 registros.


In [8]:
%%sql
SELECT title, length, release_year
FROM film
LIMIT 4;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
4 rows affected.


[('ACADEMY DINOSAUR', 86, 2006),
 ('ACE GOLDFINGER', 48, 2006),
 ('ADAPTATION HOLES', 50, 2006),
 ('AFFAIR PREJUDICE', 117, 2006)]

##### __Resultado Esperado__

**Nota:** El orden y los valores reales en los registros pueden cambiar.

| **título**         | **duración** | **año de lanzamiento** |
| ----------------- | ---------- | ---------------- |
| ACADEMY DINOSAUR  | 86         | 2006             |
| ACE GOLDFINGER    | 48         | 2006             |
| ADAPTATION HOLES  | 50         | 2006             |
| AFFAIR PREJUDICE  | 117        | 2006             |


<a id='ex03'></a>
### Ejercicio 3

Escribe una consulta SQL para obtener todas las columnas de la tabla de tiendas.


In [9]:
%%sql
SELECT *
FROM store;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
2 rows affected.


[(1, 1, 1, datetime.datetime(2006, 2, 15, 4, 57, 12)),
 (2, 2, 2, datetime.datetime(2006, 2, 15, 4, 57, 12))]

##### __Salida Esperada__ 

**Nota:** El orden y algunos detalles de los registros pueden cambiar.

| **store_id** | **manager_staff_id** | **address_id** | **last_update**         |
| ------------ | -------------------- | -------------- | ----------------------- |
| 1            | 1                    | 1              | 2006-02-15 09:57:12.000 |
| 2            | 2                    | 2              | 2006-02-15 09:57:12.000 |


<a id='3-3'></a>
### 3.3 - DONDE

La cláusula `WHERE` se utiliza para filtrar datos basados en una condición. Al final, la consulta devolverá las filas que satisfacen la condición.

```sql
SELECT *
FROM table_name
WHERE column1 = 'value1';
```


<a id='ex04'></a>
### Ejercicio 4

Escribe una consulta SQL para recuperar el nombre, apellido y correo electrónico de cada miembro del personal `activo` de la tabla `staff`.


In [10]:
%%sql 
describe  staff;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
11 rows affected.


[('staff_id', 'tinyint unsigned', 'NO', 'PRI', None, 'auto_increment'),
 ('first_name', 'varchar(45)', 'NO', '', None, ''),
 ('last_name', 'varchar(45)', 'NO', '', None, ''),
 ('address_id', 'smallint unsigned', 'NO', 'MUL', None, ''),
 ('picture', 'blob', 'YES', '', None, ''),
 ('email', 'varchar(50)', 'YES', '', None, ''),
 ('store_id', 'tinyint unsigned', 'NO', 'MUL', None, ''),
 ('active', 'tinyint(1)', 'NO', '', '1', ''),
 ('username', 'varchar(16)', 'NO', '', None, ''),
 ('password', 'varchar(40)', 'YES', '', None, ''),
 ('last_update', 'timestamp', 'NO', '', 'CURRENT_TIMESTAMP', 'DEFAULT_GENERATED on update CURRENT_TIMESTAMP')]

In [15]:
%%sql
SELECT first_name, last_name, email
FROM staff
WHERE active;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
2 rows affected.


[('Mike', 'Hillyer', 'Mike.Hillyer@sakilastaff.com'),
 ('Jon', 'Stephens', 'Jon.Stephens@sakilastaff.com')]

##### __Resultado Esperado__ 

**Nota:** El orden de los registros puede cambiar.

| **first_name** | **last_name** | **email**                    |
| -------------- | ------------- | ---------------------------- |
| Mike           | Hillyer       | Mike.Hillyer@sakilastaff.com |
| Jon            | Stephens      | Jon.Stephens@sakilastaff.com |


<a id='3-4'></a>
### 3.4 - INSERT INTO

La declaración `INSERT INTO` se utiliza para insertar nuevas filas en una tabla.

Puede insertar nuevas filas sin especificar algunas columnas, pero tendrá que escribir los nombres de las columnas y los valores que desea insertar. Esto es útil cuando algunas columnas se completan automáticamente por el valor predeterminado de la columna o cuando la columna es de tipo de datos `SERIAL`.

```sql
INSERT INTO nombre_tabla (
  columna1,
  columna2,
  columna3,
  ...
)
VALUES (
  'valor1',
  'valor2',
  'valor3',
  ...
);
```

Si está agregando una fila a una tabla y va a especificar todos los valores de las columnas, entonces no necesita especificar los nombres de las columnas y simplemente puede especificar los valores a insertar.

```sql
INSERT INTO nombre_tabla
VALUES (
  'valor1',
  'valor2',
  'valor3',
  ...
);
```


<a id='ex05'></a>
### Ejercicio 5

Escribe una consulta SQL para insertar las siguientes filas en la tabla `category_copy`:

| **category_id** | **name**  | **last_update**         |
| --------------- | --------- | ----------------------- |
| 1               | Horror    | 2006-02-15 09:46:27.000 |
| 10              | Animation | 2006-02-15 09:46:27.000 |
| 20              | Pop       | 2006-02-15 09:46:27.000 |


In [11]:
%%sql
INSERT INTO category_copy 
VALUES 
(1, 'Horror', '2006-02-15 09:46:27.000'),
(10, 'Animation', '2006-02-15 09:46:27.000'),
(20, 'Pop', '2006-02-15 09:46:27.000');

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
3 rows affected.


[]

Ejecuta la declaración `SELECT` en la tabla para verificar que los valores se agregaron correctamente:


In [12]:
%%sql
SELECT *
FROM category_copy;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
3 rows affected.


[(1, 'Horror', datetime.datetime(2006, 2, 15, 9, 46, 27)),
 (10, 'Animation', datetime.datetime(2006, 2, 15, 9, 46, 27)),
 (20, 'Pop', datetime.datetime(2006, 2, 15, 9, 46, 27))]

##### __Resultado Esperado__ 

**Nota:** El orden de los registros puede cambiar.

| **category_id** | **name**  | **last_update**         |
| --------------- | --------- | ----------------------- |
| 1               | Horror    | 2006-02-15 09:46:27.000 |
| 10              | Animación | 2006-02-15 09:46:27.000 |
| 20              | Pop       | 2006-02-15 09:46:27.000 |


<a id='3-5'></a>
### 3.5 - ACTUALIZAR

La declaración `UPDATE` se utiliza para cambiar los valores de algunas columnas en filas existentes en una tabla. Puede usar la cláusula `WHERE` para filtrar las filas que desea cambiar.

```sql
UPDATE nombre_tabla
SET
  columna2 = 'valor2',
  columna3 = 'valor3',
...
WHERE columna1 = 'valor1';
```


<a id='ex06'></a>
### Ejercicio 6

Escribe una consulta SQL para realizar los siguientes cambios en la tabla `category_copy`:

- Cambiar el valor de `last_update` a `2020-09-12 08:00:00.000` para todas las filas.
- Cambiar el valor de `category_id` a `2` para la fila con el `name` de `Animation`.
- Cambiar el valor de `name` a `Action` para la fila con el `category_id` de `1`.

Puedes agregar más celdas para cada consulta si lo deseas.


In [13]:
%%sql
UPDATE category_copy 
SET last_update = '2020-09-12 08:00:00.000';

UPDATE category_copy 
SET category_id = 2
WHERE name = 'Animation';

UPDATE category_copy 
SET name = 'Action'
WHERE category_id = 1;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
3 rows affected.
1 rows affected.
1 rows affected.


[]

Ejecuta la declaración `SELECT` en la tabla para verificar que los valores se actualizaron correctamente:


In [14]:
%%sql
SELECT *
FROM category_copy;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
3 rows affected.


[(1, 'Action', datetime.datetime(2020, 9, 12, 8, 0)),
 (2, 'Animation', datetime.datetime(2020, 9, 12, 8, 0)),
 (20, 'Pop', datetime.datetime(2020, 9, 12, 8, 0))]

##### __Salida Esperada__ 

**Nota:** El orden de los registros puede cambiar.

| **category_id** | **name**  | **last_update**         |
| --------------- | --------- | ----------------------- |
| 1               | Acción    | 2020-09-12 08:00:00.000 |
| 2               | Animación | 2020-09-12 08:00:00.000 |
| 20              | Pop       | 2020-09-12 08:00:00.000 |


<a id='3-6'></a>
### 3.6 - ELIMINAR

La instrucción `DELETE` se utiliza para eliminar filas existentes en una tabla. También se utiliza junto con la cláusula `FROM` para indicar la tabla de la cual deseas eliminar las filas. Puedes utilizar la cláusula `WHERE` para filtrar las filas que deseas eliminar.

**Debes tener cuidado porque eliminarás todas las filas de una tabla si no especificas una condición:**

```sql
DELETE FROM nombre_tabla;
```

Por eso debes agregar una condición a menos que desees eliminar todas las filas:

```sql
DELETE FROM nombre_tabla
WHERE columna1 = 'valor1';
```


<a id='ex07'></a>
### Ejercicio 7

Escribe una consulta SQL para eliminar la fila donde el `category_id` es `20` de la tabla `category_copy`.


In [15]:
%%sql
DELETE FROM category_copy 
WHERE category_id = 20;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
1 rows affected.


[]

Ejecuta la declaración `SELECT` en la tabla para verificar que las filas se eliminaron correctamente:


In [16]:
%%sql
SELECT *
FROM category_copy;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
2 rows affected.


[(1, 'Action', datetime.datetime(2020, 9, 12, 8, 0)),
 (2, 'Animation', datetime.datetime(2020, 9, 12, 8, 0))]

##### __Salida Esperada__

**Nota:** El orden de los registros puede cambiar.

| **category_id** | **name**  | **last_update**         |
| --------------- | --------- | ----------------------- |
| 1               | Acción    | 2020-09-12 08:00:00.000 |
| 2               | Animación | 2020-09-12 08:00:00.000 |


<a id='4'></a>
## 4 - Cláusulas SQL

En las siguientes secciones, verás cómo crear consultas SQL más complejas para leer datos de una base de datos utilizando las cláusulas más comunes.


<a id='4-1'></a>
### 4.1 - ALIASES

Los alias cambian temporalmente el nombre de una columna. Te permiten usar y mostrar un nombre más legible para las columnas. Para crear un alias, puedes usar la palabra clave `AS` junto a una columna para cambiar su nombre. 
Es opcional y sirve principalmente para mejorar la claridad.

```sql
SELECT
    columna1 AS Alias_Column_1,
    columna2 AS Alias_Column_2,
    columna3 AS Alias_Column_3,
    ...
FROM nombre_tabla;
```


<a id='ex08'></a>
### Ejercicio 8

Escribe una consulta SQL para obtener el título, la duración y el año de lanzamiento de la tabla `film`. Cambia los nombres de las columnas para que tengan `film_` como prefijo.


In [17]:
%%sql
SELECT 
    title AS film_title, 
    length AS film_length,
    release_year AS film_release_year
FROM film
LIMIT 10;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
10 rows affected.


[('ACADEMY DINOSAUR', 86, 2006),
 ('ACE GOLDFINGER', 48, 2006),
 ('ADAPTATION HOLES', 50, 2006),
 ('AFFAIR PREJUDICE', 117, 2006),
 ('AFRICAN EGG', 130, 2006),
 ('AGENT TRUMAN', 169, 2006),
 ('AIRPLANE SIERRA', 62, 2006),
 ('AIRPORT POLLOCK', 54, 2006),
 ('ALABAMA DEVIL', 114, 2006),
 ('ALADDIN CALENDAR', 63, 2006)]

##### __Resultado Esperado__

**Nota:** No se muestran todos los registros aquí. Su orden puede cambiar.

| **film_title**    | **film_length** | **film_release_year** |
| ----------------- | --------------- | --------------------- |
| ACADEMY DINOSAUR  | 86              | 2006                  |
| ACE GOLDFINGER    | 48              | 2006                  |
| ADAPTATION HOLES  | 50              | 2006                  |
| AFFAIR PREJUDICE  | 117             | 2006                  |
| ...               | ...             | ...                   |


<a id='4-2'></a>
### 4.2 - UNIR

La cláusula `JOIN` se utiliza para combinar los datos de múltiples tablas basándose en una columna compartida entre las tablas.

```sql
SELECT
    *
FROM table1
JOIN table2 ON table1.column1 = table2.column2;
```

Por defecto, la cláusula JOIN es equivalente a `INNER JOIN`, que combina filas de dos tablas basándose en una columna relacionada entre ellas, devolviendo solo las filas con valores coincidentes en la columna especificada en ambas tablas. También existen otros tipos de joins:
- `LEFT JOIN`: Devuelve todas las filas de la tabla izquierda y las filas coincidentes de la tabla derecha. Las filas en la tabla izquierda que no tienen coincidencias en la tabla derecha incluirán nulos para las columnas de la tabla derecha.
- `RIGHT JOIN`: Devuelve todas las filas de la tabla derecha y las filas coincidentes de la tabla izquierda. Las filas en la tabla derecha que no tienen coincidencias en la tabla izquierda incluirán nulos para las columnas de la tabla izquierda.
- `FULL JOIN`: Combina los resultados de `LEFT JOIN` y `RIGHT JOIN`, devolviendo todas las filas de ambas tablas con coincidencias cuando estén disponibles. Si no hay coincidencias, el lado faltante contendrá nulos en las columnas.


<a id='ex09'></a>
### Ejercicio 9

Escribe una consulta SQL para obtener la ciudad, dirección y distrito de cada dirección en la tabla `address`.


In [18]:
%%sql
SELECT city, address, district
FROM address
JOIN city ON address.city_id = city.city_id
LIMIT 5;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
5 rows affected.


[('A Corua (La Corua)', '939 Probolinggo Loop', 'Galicia'),
 ('Abha', '733 Mandaluyong Place', 'Asir'),
 ('Abu Dhabi', '535 Ahmadnagar Manor', 'Abu Dhabi'),
 ('Acua', '1789 Saint-Denis Parkway', 'Coahuila de Zaragoza'),
 ('Adana', '663 Baha Blanca Parkway', 'Adana')]

##### __Salida Esperada__

**Nota:** No todos los registros se muestran aquí. Su orden y valores pueden cambiar.

| **ciudad** | **dirección**         | **distrito** |
| ---------- | --------------------- | ------------ |
| Abha       | 733 Mandaluyong Place | Asir         |
| Abu Dhabi	 | 535 Ahmadnagar Manor  | Abu Dhabi    |
| ...        | ...                   | ...          |


<a id='ex10'></a>
### Ejercicio 10

Escribe una consulta SQL para obtener el país, la ciudad y la dirección de los miembros del personal activos.


In [19]:
%%sql EXPLAIN
SELECT country.country, city.city, address.address
FROM address
JOIN staff ON staff.address_id = address.address_id
JOIN city ON address.city_id = city.city_id
JOIN country ON city.country_id = country.country_id
WHERE staff.active;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
4 rows affected.


[(1, 'SIMPLE', 'staff', None, 'ALL', 'idx_fk_address_id', None, None, None, 2, 50.0, 'Using where'),
 (1, 'SIMPLE', 'address', None, 'eq_ref', 'PRIMARY,idx_fk_city_id', 'PRIMARY', '2', 'sakila.staff.address_id', 1, 100.0, None),
 (1, 'SIMPLE', 'city', None, 'eq_ref', 'PRIMARY,idx_fk_country_id', 'PRIMARY', '2', 'sakila.address.city_id', 1, 100.0, None),
 (1, 'SIMPLE', 'country', None, 'eq_ref', 'PRIMARY', 'PRIMARY', '2', 'sakila.city.country_id', 1, 100.0, None)]

**Nota:** El orden y los valores reales en los registros pueden cambiar.

| **país**   | **ciudad**  | **dirección**         |
| ----------- | ---------- | -------------------- |
| Canadá      | Lethbridge | 23 Workhaven Lane    |
| Australia   | Woodridge  | 1411 Lillydale Drive |


<a id='4-3'></a>
### 4.3 - AGRUPAR POR

La declaración SQL `GROUP BY` agrupa filas que tienen los mismos valores en columnas especificadas en filas de resumen. A menudo se utiliza con funciones de agregado (`COUNT`, `MAX`, `MIN`, `SUM`, `AVG`) para realizar cálculos en cada grupo de datos. Por lo tanto, puedes usarlo para ayudar a responder preguntas como "¿Cuál es la duración promedio de las películas por año de lanzamiento de la película?".

```sql
SELECT
    columna1,
    COUNT(columna2),
    SUM(columna3)
FROM tabla1
GROUP BY columna1;
```


<a id='ex11'></a>
### Ejercicio 11

Escribe una consulta SQL para recuperar el número de películas por clasificación en la tabla `film`.


In [20]:
%%sql
SELECT rating, COUNT(film_id) AS films_count
FROM film
GROUP BY rating;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
5 rows affected.


[('PG', 194), ('G', 178), ('NC-17', 210), ('PG-13', 223), ('R', 195)]

**Nota:** El orden y los valores reales en los registros pueden cambiar.

| **clasificación** | **cantidad_peliculas** |
| ------------------ | ---------------------- |
| PG                 | 194                    |
| G                  | 178                    |
| NC-17              | 210                    |
| PG-13              | 223                    |
| R                  | 195                    |


<a id='4-4'></a>
### 4.4 - ORDENAR POR

La cláusula `ORDER BY` se utiliza para ordenar las filas en orden ascendente o descendente basándose en una o más columnas. De forma predeterminada, esta cláusula ordenará las filas en orden ascendente, pero puedes utilizar la palabra clave `DESC` para ordenar las filas en orden descendente.

```sql
SELECT
    *
FROM table1
ORDER BY column1 DESC;
```


<a id='ex12'></a>
### Ejercicio 12

Escribe una consulta SQL para obtener el número de películas por categoría. Ordena los resultados por el número de películas en orden ascendente.


In [21]:
%%sql
SELECT category.name AS category, COUNT(film.film_id) AS films_count
FROM film
JOIN film_category ON film.film_id = film_category.film_id
JOIN category ON film_category.category_id = category.category_id
GROUP BY category.name
ORDER BY films_count;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
16 rows affected.


[('Music', 51),
 ('Horror', 56),
 ('Classics', 57),
 ('Travel', 57),
 ('Comedy', 58),
 ('Children', 60),
 ('Games', 61),
 ('Sci-Fi', 61),
 ('Drama', 62),
 ('New', 63),
 ('Action', 64),
 ('Animation', 66),
 ('Documentary', 68),
 ('Family', 69),
 ('Foreign', 73),
 ('Sports', 74)]

**Nota:** No todos los registros se muestran aquí. Su orden y valores pueden cambiar.

| **categoria** | **cantidad_peliculas** |
| -------------- | ---------------------- |
| Música         | 51                     |
| Terror         | 56                     |
| Clásicos       | 57                     |
| ...            | ...                    |


Durante este laboratorio, has escrito varias consultas SQL para manipular datos de la base de datos de muestra de alquiler de DVD. Primero, creaste consultas para recopilar información general relevante para el negocio. Luego, construiste consultas utilizando cláusulas más complejas para realizar transformaciones de datos mientras combinabas múltiples tablas. Finalmente, realizaste otras operaciones básicas de CRUD. ¡Felicidades! Ahora tienes el conocimiento básico para procesar datos utilizando SQL.


In [30]:
%%sql
DROP TABLE IF EXISTS category_copy;


 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
0 rows affected.


[]

In [32]:
%%sql
SHOW TABLES;

 * mysql+pymysql://root:***@127.0.0.1:3306/sakila
23 rows affected.


[('actor',),
 ('actor_info',),
 ('address',),
 ('category',),
 ('city',),
 ('country',),
 ('customer',),
 ('customer_list',),
 ('film',),
 ('film_actor',),
 ('film_category',),
 ('film_list',),
 ('film_text',),
 ('inventory',),
 ('language',),
 ('nicer_but_slower_film_list',),
 ('payment',),
 ('rental',),
 ('sales_by_film_category',),
 ('sales_by_store',),
 ('staff',),
 ('staff_list',),
 ('store',)]

## Referencias

<a id="1">[1]</a> Tutorial de SQL, W3schools, 2022. [En línea]. Disponible: <https://www.w3schools.com/sql/>. [Consultado: 07- Mar- 2022]
